### PR TITLE
Feature/service worker

### PR DESCRIPTION
**Pull request by julianweise**
Implemented concept of background service worker for Response-View.
ServiceWorkerConecpt base on Googles idea of progressive web apps / serviceWorkerAPI.

### CODE REVIEW

File: public/bgSyncServiceWorker.js [hunk 1]
```diff
@@ -1,5 +1,47 @@
-self.addEventListener('sync', function(event) {
-	if (event.tag === "oneTimeSync") {
-		console.log('One Sync event fired');
+function uuidv4() {
+	return ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c =>
+		(c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16)
+	)
+}
+const syncStore = {};
```
Comment (friedow) [inline public/bgSyncServiceWorker.js]: Only use global constants if you really need to.
When needed to use them, define them at the beginning of the file.

File: public/bgSyncServiceWorker.js [hunk 2]
```diff
@@ -1,5 +1,47 @@
-self.addEventListener('sync', function(event) {
-	if (event.tag === "oneTimeSync") {
-		console.log('One Sync event fired');
+function uuidv4() {
+	return ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c =>
```
Comment (friedow) [inline public/bgSyncServiceWorker.js]: What the hell?
Please provide a doc-string here.

File: public/bgSyncServiceWorker.js [hunk 3]
```diff
@@ -1,5 +1,47 @@
...
+});
+
+self.addEventListener('sync', event => {
+	// get the data by tag
+	const {type, data} = syncStore[event.tag];
+	if(type === 'response') {
+		event.waitUntil(fetch('/repsonse/', {
+			method: 'POST',
+			body: JSON.stringify(data),
+			headers: {
```
Comment (friedow) [inline public/bgSyncServiceWorker.js]: These parameters should be extracted to a constant to increase the readability.

Comment (friedow) [reply public/bgSyncServiceWorker.js]: Plus, you use them twice.

File: src/Login/Login.js [hunk 4]
```diff
@@ -2,6 +2,29 @@ import React, { Component } from 'react';
...
+	
+	handlePasswordChange(event) {
+		this.setState({password: event.target.value});
+	}
+	
+	handleEmailChange(event) {
+		this.setState({email: event.target.value});
+	}
+	
+	onSubmitLogin(event) {
```
Comment (friedow) [inline src/Login/Login.js]: Rename this to "handleSubmitLogin".

File: src/Response/Response.js [hunk 5]
```diff
@@ -1,39 +1,105 @@
...
-                        <i className="no-icon fa fa-times"></i>
-                    </button>
-                </div>
-            </div>
-        </div>
-    );
-  }
-}
+	
+	constructor(props) {
```
Comment (friedow) [inline src/Response/Response.js:6]: You don't need the props here.

File: src/Response/Response.js [hunk 6]
```diff
@@ -1,39 +1,105 @@
...
+		super(props);
+		this.onSubmitResponse = this.onSubmitResponse.bind(this);
+		this.onSubmitMistake = this.onSubmitMistake.bind(this);
+		this.state = {dataSended: false};
+	}
+
+	onSubmitMistake() {
```
Comment (friedow) [inline src/Response/Response.js]: Rename this to "handleSubmitMistake".

File: src/Response/Response.js [hunk 7]
```diff
@@ -1,39 +1,105 @@
...
+		if ('serviceWorker' in navigator) {
+			const response = {
+				authToke: this.props.authToken,
+			};
+			this.sendData('mistake', response);
+			this.setState({dataSended: false});
+		}
+	}
+	
+	onSubmitResponse(willAppear, transportation) {
```
Comment (friedow) [inline src/Response/Response.js]: Rename this to "handleSubmitResponse".

File: src/Response/Response.js [hunk 8]
```diff
@@ -1,39 +1,105 @@
...
+		super(props);
+		this.onSubmitResponse = this.onSubmitResponse.bind(this);
+		this.onSubmitMistake = this.onSubmitMistake.bind(this);
+		this.state = {dataSended: false};
+	}
+
+	onSubmitMistake() {
+		if ('serviceWorker' in navigator) {
+			const response = {
+				authToke: this.props.authToken,
```
Comment (friedow) [inline src/Response/Response.js]: Character missing "authTokeN"

File: src/Response/Response.js [hunk 9]
```diff
@@ -1,39 +1,105 @@
...
+			this.setState({dataSended: false});
+		}
+	}
+	
+	onSubmitResponse(willAppear, transportation) {
+		if ('serviceWorker' in navigator) {
+			const response = {
+				willAppear: willAppear,
+				transportation: transportation,
+				authToke: this.props.authToken,
```
Comment (friedow) [inline src/Response/Response.js]: Character missing "authTokeN"

File: src/Response/Response.js [hunk 10]
```diff
@@ -1,39 +1,105 @@
...
+				geoLocation: window.localStorage.geoLocation
+			};
+			this.sendData('response', response);
+			navigator.geolocation.getCurrentPosition(function (position) {
+				const crd = position.coords;
+				response.geolocation = {lat: crd.latitude, long: crd.longitude};
+				this.sendData('response', response);
+			}.bind(this));
```
Comment (friedow) [inline src/Response/Response.js]: Please bind "this" in the constructor.

File: src/Response/Response.js [hunk 11]
```diff
@@ -1,39 +1,105 @@
...
-        </div>
-    );
-  }
-}
+	
+	constructor(props) {
+		super(props);
+		this.onSubmitResponse = this.onSubmitResponse.bind(this);
+		this.onSubmitMistake = this.onSubmitMistake.bind(this);
+		this.state = {dataSended: false};
```
Comment (friedow) [inline src/Response/Response.js]: Grammar mistake: 

VERB | to send | sent | sent

File: src/Response/Response.js [hunk 12]
```diff
@@ -1,39 +1,105 @@
...
+				const crd = position.coords;
+				response.geolocation = {lat: crd.latitude, long: crd.longitude};
+				this.sendData('response', response);
+			}.bind(this));
+			this.setState({dataSended: true});
+		}
+	}
+	
+	sendData(type, response) {
+		navigator.serviceWorker.ready.then(function() {
```
Comment (friedow) [inline src/Response/Response.js:42]: Please use lambda functions like this () => {}.

File: src/Response/Response.js [hunk 13]
```diff
@@ -1,39 +1,105 @@
...
+	onSubmitResponse(willAppear, transportation) {
+		if ('serviceWorker' in navigator) {
+			const response = {
+				willAppear: willAppear,
+				transportation: transportation,
+				authToke: this.props.authToken,
+				geoLocation: window.localStorage.geoLocation
+			};
+			this.sendData('response', response);
+			navigator.geolocation.getCurrentPosition(function (position) {
```
Comment (friedow) [inline src/Response/Response.js]: Please use lambda functions like this () => {}.

File: src/App.js [hunk 14]
```diff
@@ -11,30 +11,27 @@ require('font-awesome/css/font-awesome.min.css');
 class App extends Component {
 	
 	constructor(props) {
+		if ('serviceWorker' in navigator) {
+			const swUrl = `${process.env.PUBLIC_URL}/bgSyncServiceWorker.js`;
+			navigator.serviceWorker.register(swUrl, {scope: '/'}).then(function(registration) {
+				console.log('ServiceWorker registration successful with scope: ', registration.scope);
+				return navigator.serviceWorker.ready;
+			}, function(err) {
```
Comment (friedow) [inline src/App.js:19]: Please use lambda functions like this () => {}.

File: src/App.js [hunk 15]
```diff
@@ -11,30 +11,27 @@ require('font-awesome/css/font-awesome.min.css');
 class App extends Component {
 	
 	constructor(props) {
+		if ('serviceWorker' in navigator) {
+			const swUrl = `${process.env.PUBLIC_URL}/bgSyncServiceWorker.js`;
+			navigator.serviceWorker.register(swUrl, {scope: '/'}).then(function(registration) {
```
Comment (friedow) [inline src/App.js:16]: Please use lambda functions like this () => {}.

File: src/config/config-template,js.js [hunk 16]
```diff
@@ -0,0 +1,4 @@
+const config = {
+	backendAPIRoute: 'http://<ip>:<port>/api/user/'
```
Comment (friedow) [inline src/config/config-template,js.js]: Please use "https://flare-resq.de/api/user" as default route.

**Review by friedow — CHANGES_REQUESTED** [review]

File: src/Response/Response.js (hunk 5 above)
Comment (julianweise) [reply src/Response/Response.js:6]: I need them to call super properly

File: src/Response/Response.js (hunk 11 above)
Comment (julianweise) [reply src/Response/Response.js]: Shame on me :/

File: src/Response/Response.js (hunk 10 above)
Comment (julianweise) [reply src/Response/Response.js]: Using arrow-lambda functions as suggested before obviates binding this context

File: src/Response/Response.js (hunk 12 above)
Comment (julianweise) [reply src/Response/Response.js:42]: Not necessary - this context is not required

File: src/App.js (hunk 15 above)
Comment (julianweise) [reply src/App.js:16]: Not necessary - this context is not required

**Review by friedow — APPROVED** [review]

**Pull request merged**